File: src/Tokeniser.py
#! /usr/bin/env python

from ply import lex


class Lexer:
    UNIQUE_WORDS = {
        }

    tokens = [
        'WORD',
        'OTHER',
        ]

    # token regexes
    t_ignore = " \t"

    # track linenos
    def t_newline(self, t):
        r'\n+'
        t.lexer.lineno += len(t.value)

    # skip errors
    def t_error(self, t):
        print "illegal sequence '%s'" % t.value[0]
        t.lexer.skip(1)

    # handle words
    def t_WORD(self, t):
        r'(?i)[a-z\-_0-9(){}:,]+'
        t.type = self.UNIQUE_WORDS.get(t.value, 'WORD')
        return t

    def t_OTHER(self, t):
        r'.'
        return t

    def __init__(self, **kwargs):
        self.lexer = lex.lex(module=self, **kwargs)

    def set_input(self, input_data):
        self.lexer.input(input_data)

    def lexdata(self):
        return self.lexer.lexdata


if __name__ == '__main__':
    l = Lexer()
    l.set_input('this is a TEST but not a simple test: 2,2-bis(4-hydroxy-3-tert-butylphenyl)propane')
    for tok in l.lexer:
        print tok

File: R/Tokeniser.R
library(rly)

TOKENS = c('WORD', 'OTHER')
LITERALS = c('=','+','-','*','/', '(',')')

Lexer <- R6::R6Class("Lexer",
  public = list(
    tokens = TOKENS,
    literals = LITERALS,
    t_WORD = function(re='(?i)[a-z0-9]+', t) {
      return(t)  
    },
    t_OTHER = function(re='.', t) {
      return(t)
    },
    t_ignore = " \t",
    t_newline = function(re='\\n+', t) {
      t$lexer$lineno <- t$lexer$lineno + nchar(t$value)
      return(NULL)
    },
    t_error = function(t) {
      cat(sprintf("Illegal character '%s'", t$value[1]))
      t$lexer$skip(1)
      return(t)
    }
  )
)

lexer  <- rly::lex(Lexer)

lexer$input('this is a TEST, possibly a simplistic test, but not a SIMPLE test because it contains : 2,2-bis(4-hydroxy-3-tert-butylphenyl)propane and 2,2-bis(4-tert-3-hydroxy-butylphenyl)propane amongst other things')

while (lexer$lexpos <= lexer$lexlen) {
  print(lexer$token())
}

lexer$input('this is a TEST, possibly a simplistic test, but not a SIMPLE test because it contains:\n2,2-bis(4-hydroxy-3-tert-butylphenyl)propane and 2,2-bis(4-tert-3-hydroxy-butylphenyl)propane amongst other things')

while (lexer$lexpos <= lexer$lexlen) {
  tok <- lexer$token()
  print(tok$type)
  print(tok$value)
  print(tok$lexpos)
  print(tok)
}
